File: src/com/suushiemaniac/cubing/bld/util/ClosureUtil.java
package com.suushiemaniac.cubing.bld.util;

import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;

public class ClosureUtil {
	public static <T, U, R> Function<U, R> curry(BiFunction<T, U, R> unBound, T arg) {
		return fnArg -> unBound.apply(arg, fnArg);
	}

	public static <T, U, R> Function<T, R> curryRight(BiFunction<T, U, R> unBound, U arg) {
		return fnArg -> unBound.apply(fnArg, arg);
	}

	public static <T, R> Supplier<R> curry(Function<T, R> unBound, T arg) {
		return () -> unBound.apply(arg);
	}

	public static <T> Predicate<T> predicatize(Function<T, Boolean> origin) {
		return origin::apply;
	}

	public static <T, U, R> BiFunction<U, T, R> reversingArguments(BiFunction<T, U, R> original) {
		return (a, b) -> original.apply(b, a);
	}

	public static <T> Supplier<T> constant(T value) {
		return () -> value;
	}

	public static <T, R> Function<T, R> always(R value) {
		return (in) -> value;
	}

	public static <T, R> Function<T, R> always(Supplier<R> value) {
		return (in) -> value.get();
	}

	public static <T, R> Function<T, R> tryCatch(Function<T, R> throwing) {
		return (t) -> {
			try {
				return throwing.apply(t);
			} catch (Exception e) {
				throw new RuntimeException(e);
			}
		};
	}
}

File: src/com/suushiemaniac/cubing/bld/util/ClosureUtil.kt
package com.suushiemaniac.cubing.bld.util

object ClosureUtil {
    fun <T, U, R> ((T, U) -> R).curry(arg: T): (U) -> R {
        return { this(arg, it) }
    }

    fun <T, U, R> ((T, U) -> R).curryRight(arg: U): (T) -> R {
        return { this(it, arg) }
    }

    fun <T, R> ((T) -> R).curry(arg: T): () -> R {
        return { this(arg) }
    }

    fun <T, U, R> ((T, U) -> R).reversingArguments(): (U, T) -> R {
        return { a, b -> this(b, a) }
    }

    fun <T> T.constant(): () -> T {
        return { this }
    }

    fun <T, R> R.always(): (T) -> R {
        return { this }
    }

    fun <T, R> (() -> R).always(): (T) -> R {
        return { this() }
    }

    fun <T, R> ((T) -> R).tryCatch(): (T) -> R {
        return {
            try {
                this(it)
            } catch (e: Exception) {
                throw RuntimeException(e)
            }
        }
    }
}